src/bvals/comms/bvals_utils.hpp

```diff
@@ -35,22 +35,35 @@
 #include "utils/loop_utils.hpp"
 
 namespace parthenon {
-inline std::tuple<int, int, std::string, int>
-SendKey(const MeshBlock *pmb, const NeighborBlock &nb,
-        const std::shared_ptr<Variable<Real>> &pcv) {
+inline auto &GetSenderGid(Mesh::channel_key_t &key) { return std::get<0>(key); }
+inline auto &GetReceiverGid(Mesh::channel_key_t &key) { return std::get<1>(key); }
+inline auto &GetVariable(Mesh::channel_key_t &key) { return std::get<2>(key); }
+inline auto &GetLocIdx(Mesh::channel_key_t &key) { return std::get<3>(key); }
+
+inline Mesh::channel_key_t SendKey(const MeshBlock *pmb, const NeighborBlock &nb,
+                                   const std::shared_ptr<Variable<Real>> &pcv,
+                                   BoundaryType btype) {
   const int sender_id = pmb->gid;
   const int receiver_id = nb.gid;
   const int location_idx = nb.offsets.GetIdx();
-  return {sender_id, receiver_id, pcv->label(), location_idx};
+  int other = (nb.gid == pmb->gid && (btype == BoundaryType::gmg_restrict_recv ||
+                                      btype == BoundaryType::gmg_restrict_send))
+                  ? 1
+                  : 0;
+  return {sender_id, receiver_id, pcv->label(), location_idx, other};
```

Comment:
Add more information to the communication key so that we can represent same-to-same restriction and same-to-same prolongation as different channels.